asllib/doc/ASLSemanticsReference.tex

```diff
@@ -19,160 +283,1634 @@
 
 \chapter{Preamble}
 
-The semantics of ASL are embodied in a function that takes as input an ASL program
-and returns all possible behaviours of that program.
+The semantics of ASL defines all valid behaviors of a given ASL specification.
+More precisely, an ASL specification is first parsed into an \emph{abstract syntax tree},
+or AST, for short. Second, a type checker analyzes the \emph{parsed AST} for correctness and, if successful,
+returns a \emph{static environment} and a \emph{typed AST}. Finally, tools such as interpreters and verifiers
+can operate over the typed AST, based on the definition of the semantics, to test and analyze a given specification.
+
+The abstract syntax, parsed AST, and typed AST are defined in the abstract syntax reference~\cite{ASLAbstractSyntaxReference}.
+The ASL type system, which defines the set of parsed ASTs that are considered valid and how to
+produce the corresponding typed ASTs and static environments is defined in the typing reference~\cite{ASLTypingReference}.
+
+\paragraph{Basic Semantic Concepts.}
+The ASL semantics is given by a set of \emph{semantic relations} between \emph{configurations}.
+Intuitively, configurations represent intermediate states during the execution of a program and they encapsulate
+the information needed to ``transition'' into other configurations.
+Configurations can include a \emph{dynamic environment}, which binds variables to values;
+the typed AST node that needs to be ``evaluated'';
+a \herd\ \emph{execution graph}; and possibly other pieces of information, for example intermediate expression values.
+The semantic relations are constructively defined via \emph{semantic rules}.
+The rules, in turn, inductively define transitions from configurations with an AST node
+in terms of transitions starting from configurations with its child nodes.
+
+\paragraph{Execution.}
+A valid execution of an ASL specification transitions from an \emph{initial configuration},
+which consists of the given specification and the standard library specification, to an output configuration
+consisting of an output value and a \herd\ execution graph.
+
+\paragraph{Primitive Subprograms.}
+The semantics of ASL is parameterized by a set of primitive subprograms ---
+subprograms whose implementation is not defined by an ASL body and whose effect on the dynamic environment
+is defined externally. Critically, access to memory is given by primitive subprograms.
+
+We define two types of semantics --- a \emph{sequential semantics} and a \emph{concurrent semantics}.
+
+\paragraph{Concurrent Semantics.}
+The concurrent semantics includes \herd\ execution graphs.
+Intuitively, these graphs define read and write effects to variables and constraints over those effects.
+Together with the constraints that define the ARM memory model~\cite{AlglaveDGHM21}, these graphs axiomatically define
+the valid sequences of read and write operations induced by a given specification.
```

Comment:
valid sequences of read and write operations induced by a given specification -> valid interactions of shared variables of a given specification